flows/msg.go

```diff
@@ -178,16 +181,56 @@ type TemplatingComponent struct {
 	Params []TemplatingParam `json:"params"`
 }
 
+var templateVariableRegex = regexp.MustCompile(`{{(\d+)}}`)
+
+func getTemplateVariableCount(s string) int {
+	count := 0
+	for _, m := range templateVariableRegex.FindAllStringSubmatch(s, -1) {
+		if v, _ := strconv.Atoi(m[1]); v > count {
+			count = v
+		}
+	}
+	return count
+}
+
+// Preview returns the content and display for given template component using these templating params
+func (tc *TemplatingComponent) Preview(c assets.TemplateComponent) (string, string) {
+	content := c.Content()
+	display := c.Display()
+	numContentParams := getTemplateVariableCount(content)
+	numDisplayParams := getTemplateVariableCount(display)
+
+	// replace {{?}} placeholders in component content
+	for i := 0; i < numContentParams; i++ {
+		value := ""
+		if i < len(tc.Params) {
+			value = tc.Params[i].Value
+		}
+		content = strings.ReplaceAll(content, fmt.Sprintf("{{%d}}", i+1), value)
+	}
+
+	// replace {{?}} placeholders in component display using any remaining param values
+	for i := 0; i < numDisplayParams; i++ {
+		value := ""
+		if (numContentParams + i) < len(tc.Params) {
+			value = tc.Params[numContentParams+i].Value
+		}
+		display = strings.ReplaceAll(display, fmt.Sprintf("{{%d}}", i+1), value)
+	}
+
+	return content, display
```

Comment:
@norkans7 @ericnewcomer the complexity of this has me wondering if it's a good idea to put variables for the content part of a component and variables for the display part in the same list... because it probably means both courier and the editor would need the same logic to know which variables belong to which.

According to META's woeful documentation, for a URL type button, it can have a variable in the URL and a variable in the text:

<img width="837" alt="Captura de pantalla 2024-03-20 a la(s) 14 18 35" src="https://github.com/nyaruka/goflow/assets/675558/942fe1cb-c6d6-4733-8d35-d307f0e6202c">

But I can't find any examples of how to pass variables for both parts...

Reply:
One alternative would be to add `display_params` as separate field to components

Reply:
So far they can not have the variable at the same time, we have display for URL type and in that case the display cannot have variables, the variable will be in the URL only

in other words, a display cannot have variables, if we have the quick_reply type the variable will in the content itself we won't be having any display for that type

Reply:
Or just not support URL buttons with variables in the display part... I hate making this so complicated for a thing nobody might use.

Reply:
Meta already do not support that now, we can add that support when they add that too.
buttons can have maximum one variable if it is the URL type the variable is for the URL part only.

Reply:
Ugh so their docs are just wrong.. guess I'll strip out that complexity then